pkg/controller/master/virtualmachine/vmi_controller.go

```diff
@@ -35,6 +41,71 @@ type VMIController struct {
 	recorder            record.EventRecorder
 }
 
+// SyncHarvesterVMILabelsToPod ensures that all Harvester labels (i.e. those
+// with the harvesterhci.io/ prefix) from the VirtualMachineInstance are synced
+// to the Pod
+func (h *VMIController) SyncHarvesterVMILabelsToPod(_ string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmi == nil || vmi.DeletionTimestamp != nil {
+		return vmi, nil
+	}
+
+	logrus.Debugf("Syncing labels %v for VMI %v to Pod", vmi.Labels, vmi.Name)
+
+	harvesterVMILabels := map[string]string{}
+	for label := range vmi.Labels {
+		if strings.HasPrefix(label, HarvesterLabelPrefix) {
+			harvesterVMILabels[label] = vmi.Labels[label]
+		}
+	}
+
+	activePods := vmi.Status.ActivePods
+	if len(activePods) < 1 {
+		logrus.Debugf("VMI %v does not have active Pods", vmi.Name)
+		return vmi, nil
+	}
+
+	nodeName := vmi.Status.NodeName
+	if nodeName == "" {
+		logrus.Debugf("Could not identify which node VMI %v runs on", vmi.Name)
+		return vmi, nil
+	}
+
+	pods, err := h.podCache.GetByIndex(indexeres.PodByNodeNameIndex, nodeName)
+	if err != nil {
+		return vmi, fmt.Errorf("failed to find pods for VMI %v, %v", vmi.Name, err)
+	}
+
+	var toUpdate = make([]*corev1.Pod, 0)
+	for _, pod := range pods {
+		for active := range activePods {
+			if pod.UID == active {
+				toUpdate = append(toUpdate, pod.DeepCopy())
+			}
+		}
+	}
+
+	for _, pod := range toUpdate {
+		// delete Harvester labels from Pod, if they are deleted from VMI
+		for podLabel := range pod.Labels {
+			_, ok := harvesterVMILabels[podLabel]
+			if strings.HasPrefix(podLabel, HarvesterLabelPrefix) && !ok {
+				delete(pod.Labels, podLabel)
+			}
+		}
+
+		// copy labels from VMI to Pod
+		for label := range harvesterVMILabels {
+			pod.Labels[label] = harvesterVMILabels[label]
+		}
+
+		_, err := h.podClient.Update(pod)
```

Comment:
vmi onChange happens many times, use deepEqual to check if pod is really changed then call update will be faster

the deepEqual runs in local go thread; the .Update runs on apiServer, the latter is much slower

Reply:
I've made several changes that I think should improve the performance of this entire method.
Avoiding the `Update` as much as possible is one of them, but the `DeepCopy` also now only happens when the `Update` happens.
Additionally, I've replaced a few of the for-loops with functions from the standard library and I've gotten rid of the dynamically growing array.
The last big improvement in this department is that I'm now using a different Pod indexer, which only returns Pods associated with the VirtualMachine, not all Pods running on the node where the VirtualMachineInstance is running, so there are a lot less Pods to look through.